### DISCUSSION TITLE
Get audio output directly

### DISCUSSION THREAD

**Discussion by jhoelzl** [pndurette/gTTS#376] [Q&A]
Hello,

i have following code:

```
text = "Hello, my name is Josef."
targetLanguage = 'de'
tts = gTTS(text, targetLanguage)  
tts.save("synthesized.mp3")
```

Since i play the output directly after this code, is there a way to get the audio output directly as a variable? So i do not have to save and open the mp3 file again.

Thanks!
Josef

**Comment by pndurette** [comment]
Yes it's possible. You need a module that will play from a file-like/file object. Instead of doing `tts.save()` you can use `tts.write_to_fp()` to write to a file-like object that you can then reuse to play. As a really simple example:

```
from gtts import gTTS
from tempfile import TemporaryFile
tts = gTTS(text='Hello', lang='en')
f = TemporaryFile()
tts.write_to_fp(f)
# Play f
f.close()
```

**Comment by panev** [comment]
This would indeed be fantastic. 

I'm having issues with finding a decent approach. Most audio playback libraries are hectic to use in this scanario.

**Comment by jhoelzl** [comment]
@pndurette thanks for help! Before, i used

`os.system('cvlc --play-and-exit ' + "synthesized.mp3")`

to play the audio file. Now, i have a an audio object called `f`.

I tried 
`os.system('cvlc --play-and-exit ' + f)`

but is does not work. Do you have another example how to playback the generated audio object?

Thanks for helping!

Josef

**Comment by anselm94** [comment]
Try [Pyglet](https://bitbucket.org/pyglet/pyglet/wiki/Home) - a multimedia and windowing library for Python (~3MB)

```
from gtts import gTTS
from time import sleep
import os
import pyglet

tts = gTTS(text='Hello World', lang='en')
filename = '/tmp/temp.mp3'
tts.save(filename)

music = pyglet.media.load(filename, streaming=False)
music.play()

sleep(music.duration) #prevent from killing
os.remove(filename) #remove temperory file
```

**Comment by jhoelzl** [comment]
Thanks @anselm94, but i want to avoid saving the audio data on the file system. Instead, i directly want to play the audio output from module gTTS. Do you think this is possible with pyglet?

**Comment by anselm94** [comment]
Currently [Pyglet](http://www.pyglet.org) doesn't support using file-like objects. As the project documentation [here](https://pyglet.readthedocs.io/en/pyglet-1.2-maintenance/programming_guide/media.html#loading-media) (see **Loading Media** section) says

>  Future versions of pyglet will also support reading from arbitrary file-like objects, however a valid filename must currently be given.

**Comment by rishabhrkaushik** [comment]
@jhoelzl 
Having the same issue. Have you found the solution yet.

**Comment by jhoelzl** [comment]
@rishabhrkaushik unfortunately not.

**Comment by Khoa-NT** [comment]
Hi, are there any ways to output file.wav , not mp3 ? i don't know how to use python play mp3 file.

The pyglet like anselm94 said can't open file .mp3

thank you and regards

**Comment by common83** [comment]
@shaolinkhoa  using pyMedia lib
here example:
https://github.com/common83/playmp3/blob/master/tts_play.py

**Comment by AllenAkhaumere** [comment]
Use pygame multimedia package

**Comment by AllenAkhaumere** [comment]
pygame.mixer.init()
pygame.mixer.music.load(file.mp3)
pygame.mixer.music.play() 
 It works in real-time. Good luck

**Comment by yagiziskirik** [comment]
My solution to unicode-charactered languages:
```
#!/usr/bin/env python
#-*- coding: utf-8 -*-

from gtts import gTTS
import os

class Speech(object):
	def __init__(self):
		while True:
			try:
				a = raw_input("What you want me to say: ")
				a = unicode(a, "utf-8")
				tts = gTTS(text=a, lang="en")
				
				testfile = "/tmp/temp.mp3"
				tts.save(testfile)

				os.system("mpg123 /tmp/temp.mp3")
				os.system("clear")
				os.system("rm %s" %(testfile))
			except UnicodeDecodeError:
				print "Some characters are not supported."
spc = Speech()
```
You have to have a mpg123 to run this program, but it deletes the ".mp3" file everytime it finishes the phrase.

**Comment by VeNoMouS** [comment]
@yagiz190799 

jesus... `rm` via `os.system()` really??? `os.unlink()` dude!

Also your clear can be safer with ```print "\033[H\033[J"``` 

Might i also recommend pep8 standard, your using tabs for indentation (smh)... spaces mate!!

some people just want to see the world burn :P

**Comment by patel33hardik** [comment]
If are you using a web browser,
-save .mp3 to the temp folder on the server
-return file path or name 
- use jQuery and HTML 
   `var file_name = response['Data'];
                var audio = new Audio();
                audio.src = "frontend/temp_mp3/"+file_name;
                audio.addEventListener("loadedmetadata", function(_event) {
                    var duration = (audio.duration * 1000);
                    audio.play();
                    $("#TextToAudio").attr("src","frontend/images/text_to_audio_icon.gif");
                    deleteAudioFile(file_name)
                    setTimeout(function(){
                        document.getElementById('TextToAudio').disabled=false;
                        $("#TextToAudio").attr("src","frontend/images/text_to_audio_icon.png");
                    }, duration);
                });`
- after content download, delete file from the server using ajax request.

**Comment by so-germaine** [comment]
@pndurette Hello, I'm having trouble finding a python module that plays file-like objects. What such modules have you used in the past with gTTS? Thank you.

**Comment by VeNoMouS** [comment]
@so-germaine  you could do something like...

```
import pygame, StringIO

def gm_PlaySong(self, url):
    req = urllib2.Request(url)
    resp = urllib2.urlopen(req)
    song = StringIO.StringIO(resp.read()) 

    pygame.mixer.music.load(song)
```

**Comment by sdsouza33** [comment]
How to play the file from BytesIO?
I have used 
file=BytesIO () 
Tts.write_to_fp(file)
Now, what should I do? 
I want to play this file 
None of the methods above worked for me.

**Comment by VeNoMouS** [comment]
@sdsouza33 `file.seek(0)`  before writing .. but that said... what is in `file`? from you're example there is nothing there... you're just declaring a pointer...

**Comment by bghanchi** [comment]
> Yes it's possible. You need a module that will play from a file-like/file object. Instead of doing `tts.save()` you can use `tts.write_to_fp()` to write to a file-like object that you can then reuse to play. As a really simple example:
> 
> ```
> from gtts import gTTS
> from tempfile import TemporaryFile
> tts = gTTS(text='Hello', lang='en')
> f = TemporaryFile()
> tts.write_to_fp(f)
> # Play f
> f.close()
> ```

It's not working for me

**Comment by jbrepogmailcom** [comment]
Hello! Seems this issue did not offer generic solution and yet it has been closed. Pygame does not install on my OrangePi Zero, so I can not play in python from file-like object. Or is there any other solution?
Thanks

**Comment by hpolloni** [comment]
This worked for me:
```python
from gtts import gTTS
import pygame
from io import BytesIO

pygame.init()

def say(text):
    tts = gTTS(text=text, lang='en')
    fp = BytesIO()
    tts.write_to_fp(fp)
    fp.seek(0)
    pygame.mixer.init()
    pygame.mixer.music.load(fp)
    pygame.mixer.music.play()
    while pygame.mixer.music.get_busy():
        pygame.time.Clock().tick(10)
```

**Comment by harish6471** [comment]
Try this


```import speech_recognition as sr
from gtts import gTTS
import os
import time
import playsound

def speak(text):
    tts = gTTS(text=text, lang='en',slow=False)
    filename ='speech.mp3'
    tts.save(filename)
    playsound.playsound(filename)
    speak("hello ")```

**Comment by pouyaSamie** [comment]
@hpolloni solution did the job just with one change 
removing pygame.init() line and the rest was ok.

**Comment by SmellyArmure** [comment]
Hello everybody,

This (see below) worked for me, the only difficulty was to find : 
- how to use the file-like object and 
- which module to use to read it.

> > Yes it's possible. You need a module that will play from a file-like/file object. Instead of doing `tts.save()` you can use `tts.write_to_fp()` to write to a file-like object that you can then reuse to play. As a really simple example:
> > ```
> > from gtts import gTTS
> > from tempfile import TemporaryFile
> > tts = gTTS(text='Hello', lang='en')
> > f = TemporaryFile()
> > tts.write_to_fp(f)
> > # Play f
> > f.close()
> > ```
> 
> It's not working for me


Here is the shorter solution I found :

```
import time
from gtts import gTTS
from tempfile import TemporaryFile
from IPython.display import Audio

tts = gTTS(text='Dataset successfully imported', lang='en')
f = TemporaryFile()
tts.write_to_fp(f)
f.seek(0)
Audio(f.read(), autoplay=True)
# time.sleep(5) # doesn't work to avoid the file to close to fast...
# f.close() # ... this has to be in the next cell otherwise the sound is not played

```

The only difficulty is that I did not manage to close the file-like object right away after the playing command.
I tryed to put the program to sleep 5 seconds with time.sleep(5) but without success, or to close in the next cell of notebook.... Still interested if anyone knows.

**Comment by CIArjun** [comment]
Thanks a lot @SmellyArmure ...

Your code worked for me... 
And for others of above all solution his solution solves the issue, please make use of it!

**Comment by SmellyArmure** [comment]
You're welcome.
A little update.  Here is the final version I use, it works quite well and helps keeping track of the execution of a long notebook while doing something else on the PC :

```
from IPython.display import Audio
from gtts import gTTS
from tempfile import TemporaryFile

# To play audio text-to-speech during execution
def speak(my_text):
  f = TemporaryFile()
  tts = gTTS(text=my_text, lang='en')
  tts.write_to_fp(f)
  f.seek(0)
  return Audio(f.read(), autoplay=True)
  f.close()

speak('checkpoint number one : Packages and modules successfully imported')
```

**Comment by VeNoMouS** [comment]
`f.close()` will never be called

```python
from IPython.display import Audio
from gtts import gTTS
from tempfile import TemporaryFile

# To play audio text-to-speech during execution
def speak(my_text):
    with TemporaryFile() as f:
        gTTS(text=my_text, lang='en').write_to_fp(f)
        f.seek(0)
        return Audio(f.read(), autoplay=True)
```

would be alot cleaner...

however

```python
import io
from IPython.display import Audio
from gtts import gTTS

# To play audio text-to-speech during execution
def speak(my_text):
    with io.BytesIO() as f:
        gTTS(text=my_text, lang='en').write_to_fp(f)
        f.seek(0)
        return Audio(f.read(), autoplay=True)
```

would be a better way to go, and simply store the gtts in memory rather than writing to an actual file.

**Comment by jackpeters667** [comment]
![Screenshot from 2020-04-01 23-05-01](https://user-images.githubusercontent.com/47973728/78186600-48ceef80-746d-11ea-96b9-aaaf5f13ba39.png)
@VeNoMouS this is gives me no output. What am I missing?

**Comment by VeNoMouS** [comment]
Sorry I don't actually have IPython, it could be it requires `f` rather than `f.read()`... reading the doco at https://ipython.org/ipython-doc/dev/api/generated/IPython.display.html it looks like it can take raw data, if this is the case, it should simply be `f` not `f.read()`

for those without IPython this plays fine for me..

```python

import io
import pygame
from gtts import gTTS

# To play audio text-to-speech during execution
def speak(my_text):
    with io.BytesIO() as f:
        gTTS(text=my_text, lang='en').write_to_fp(f)
        f.seek(0)
        pygame.mixer.init()
        pygame.mixer.music.load(f)
        pygame.mixer.music.play()
        while pygame.mixer.music.get_busy():
            continue

speak('luke, i am your father')
```

**Reply by NooRMaseR** [reply]
Is there's another way to play the audio except the pygame ?, because I'm using flask and i want to send that auido generated to JavaScript, i was able to send files but i is there's a way to send the voice directly to a server ?

**Comment by CIArjun** [comment]
Thanks a lot @pndurette 😊 will definitely use it

**Comment by SmellyArmure** [comment]
> ![Screenshot from 2020-04-01 23-05-01](https://user-images.githubusercontent.com/47973728/78186600-48ceef80-746d-11ea-96b9-aaaf5f13ba39.png)
> @VeNoMouS this is gives me no output. What am I missing?

I confirm that this exact version given by VeNoMous works fine for me (I'm on Google Colab).
@jackpeters667 No idea what the problem might be...

**Comment by linubajy** [comment]
I want to overwrite an audio file that reads out answers every time a qn is asked.I have used HTML files and gtts module
How do I overwrite (in python)the audio, for each time ,so that it can be called from an html page.Obviously an object can't be called in html page so I require it to name in MP3 format each time

**Comment by Markandeshwar** [comment]
> Try [Pyglet](https://bitbucket.org/pyglet/pyglet/wiki/Home) - a multimedia and windowing library for Python (~3MB)
> 
> ```
> from gtts import gTTS
> from time import sleep
> import os
> import pyglet
> 
> tts = gTTS(text='Hello World', lang='en')
> filename = '/tmp/temp.mp3'
> tts.save(filename)
> 
> music = pyglet.media.load(filename, streaming=False)
> music.play()
> 
> sleep(music.duration) #prevent from killing
> os.remove(filename) #remove temperory file
> ```

if u put this in loop then its showing the following error. kindly let me know how  to debug this.
Traceback (most recent call last):
  File "playsound2.py", line 37, in <module>
    music = pyglet.media.load("ThankYou.mp3", streaming=True)
  File "C:\Users\Pleasure\AppData\Local\Programs\Python\Python38-32\lib\site-pac
kages\pyglet\media\__init__.py", line 130, in load
    loaded_source = decoder.decode(file, filename, streaming)
  File "C:\Users\Pleasure\AppData\Local\Programs\Python\Python38-32\lib\site-pac
kages\pyglet\media\codecs\wave.py", line 106, in decode
    return WaveSource(filename, file)
  File "C:\Users\Pleasure\AppData\Local\Programs\Python\Python38-32\lib\site-pac
kages\pyglet\media\codecs\wave.py", line 53, in __init__
    file = open(filename, 'rb')
PermissionError: [Errno 13] Permission denied: 'ThankYou.mp3'
Exception ignored in: <function WaveSource.__del__ at 0x039DCAD8>
Traceback (most recent call last):
  File "C:\Users\Pleasure\AppData\Local\Programs\Python\Python38-32\lib\site-pac
kages\pyglet\media\codecs\wave.py", line 76, in __del__
    self._file.close()
AttributeError: 'WaveSource' object has no attribute '_file'

**Comment by VeNoMouS** [comment]
Do a with for pyglet it’s holding the file open to write

**Comment by themdrizwanansari** [comment]
guys use **playsound()** with **NamedTemporaryFile()**

it works but I fail to see significant decrease in delay in output.

**Comment by JoergZ2** [comment]
Is there a running solution for playing TTS output "on demand"? Writing to a file and playing back by any audioplayer works fine. But no playing the file-like object. I've tested pygame, vlc, audioplayer no one can play the gTTS.write_to_fp() object. But I don't want to write a file because it's on a sd-card.

**Comment by sijah** [comment]
> This worked for me:
> 
> ```python
> from gtts import gTTS
> import pygame
> from io import BytesIO
> 
> pygame.init()
> 
> def say(text):
>     tts = gTTS(text=text, lang='en')
>     fp = BytesIO()
>     tts.write_to_fp(fp)
>     fp.seek(0)
>     pygame.mixer.init()
>     pygame.mixer.music.load(fp)
>     pygame.mixer.music.play()
>     while pygame.mixer.music.get_busy():
>         pygame.time.Clock().tick(10)
> ```

this works for me

**Comment by JoergZ2** [comment]
I know the code but it throws errors:
`
Traceback (most recent call last):
  File "/home/jz/Skripte/TTS/googletts-4.py", line 19, in <module>
    say(text)
  File "/home/jz/Skripte/TTS/googletts-4.py", line 13, in say
    pygame.mixer.music.load(fp)
pygame.error: Module format not recognized
`
I only added this two lines:
`text = "This is an example"
say(text)
`
If I stored the gTTS output in a file e. g. tts.save('my_file.mp3') and used pygame.mixer.music.load('my_file.mp3') it works. Same result with other audioplayers like python-vlc or audioplayer.AudioPlayer(). Ther must be a difference between the file-like object and a "real" file.

**Comment by mlazzarotto** [comment]
> This worked for me:
> 
> ```python
> from gtts import gTTS
> import pygame
> from io import BytesIO
> 
> pygame.init()
> 
> def say(text):
>     tts = gTTS(text=text, lang='en')
>     fp = BytesIO()
>     tts.write_to_fp(fp)
>     fp.seek(0)
>     pygame.mixer.init()
>     pygame.mixer.music.load(fp)
>     pygame.mixer.music.play()
>     while pygame.mixer.music.get_busy():
>         pygame.time.Clock().tick(10)
> ```

I'm trying to run the above code on Python 3.9.1, but everytime I run it I get the following error.  
`pygame.error: ModPlug_Load failed`  
Can someone help me?

**Comment by dinda46** [comment]
i use this code and the output says "Hello 123". it automatically played the audio and if you see in your folder there will be no "audio1.mp3" file

```python
import os
from gtts import gTTS
import pygame

def speak(text):
	tts=gTTS(text="Hello 123", lang='en')
	filename="audio1.mp3"
	tts.save(filename)
	pygame.mixer.init()
	pygame.mixer.music.load("audio1.mp3")
	pygame.mixer.music.play()
	while pygame.mixer.music.get_busy():
		pygame.time.Clock().tick(10)
	pygame.mixer.quit()
	os.remove(filename)
```

**Comment by themdrizwanansari** [comment]
> guys use **playsound()** with **NamedTemporaryFile()** from tempfile
> 
> it works but I fail to see significant decrease in delay in output.

If **pygame** is not a necessity you all can try to use this solution.

**Comment by sorenmulli** [comment]
Hello everyone, I would like to share this solution using `pyglet` which was the only nice way I could resolve this without writing to disk:

```python
import io
import time

import pyglet
from gtts import gTTS

pyglet.options["audio"] = ("pulse",)
TICK = .1

def speak(words: str, lang: str="en"):
    with io.BytesIO() as f:
        gTTS(text=words, lang=lang).write_to_fp(f)
        f.seek(0)
        sound = pyglet.media.load("_.mp3", file=f)
    player = sound.play()
    while player.playing:
        pyglet.app.platform_event_loop.dispatch_posted_events()
        pyglet.clock.tick()
        time.sleep(TICK)
``` 

Edit: Improved using suggestions by @VeNoMouS

**Comment by VeNoMouS** [comment]
> Hello everyone, I would like to share this solution using `pyglet` which was the only nice way I could resolve this without writing to disk:
> 
> ```python
> import io
> import time
> 
> import pyglet
> from gtts import gTTS
> 
> pyglet.options["audio"] = ("pulse",)
> TICK = 0.2
> 
> def speak(words: str, lang: str="en"):
>     with io.BytesIO() as f:
>         gTTS(text=words, lang=lang).write_to_fp(f)
>         f.seek(0)
>         sound = pyglet.media.load("_.mp3", file=f)
>     player = sound.play()
>     playtime = player.source.duration
>     while (playtime - TICK) > -
>         pyglet.app.platform_event_loop.dispatch_posted_events()
>         pyglet.clock.tick()
>         playtime -= TICK
>         time.sleep(TICK)
> ```

Would it not make more sense to do the following..

```python
import io
import time

import pyglet
from gtts import gTTS

pyglet.options["audio"] = ("pulse",)

def speak(words: str, lang: str="en"):
    with io.BytesIO() as f:
        gTTS(text=words, lang=lang).write_to_fp(f)
        f.seek(0)
        
        player = pyglet.media.load('_.mp3', file=f).play()
        while player.source.duration != player.time:
            pyglet.app.platform_event_loop.dispatch_posted_events()
            pyglet.clock.tick()
```

Or better yet...

```python
import io
import time

import pyglet
from gtts import gTTS

pyglet.options["audio"] = ("pulse",)

def speak(words: str, lang: str="en"):
    with io.BytesIO() as f:
        gTTS(text=words, lang=lang).write_to_fp(f)
        f.seek(0)
        
        player = pyglet.media.load('_.mp3', file=f).play()
        while player.playing:
            pyglet.app.platform_event_loop.dispatch_posted_events()
            pyglet.clock.tick()
```

**Comment by sorenmulli** [comment]
@VeNoMouS, that definitely looks nicer, but I can't get it to work: I only get the first small snippet of sound and only when i use `dispatch_posted_events()`, I get a satisfying result.

**Comment by VeNoMouS** [comment]
@sorenmulli updated those examples for you then ;P

**Comment by SG-SakethGupta** [comment]
Hey guys im trying to do a voice assistant 

HELP

im trying to make the python file speak what i actually want it to speak. I have tried play sound and the temporary files. NO PIGLET or pygame plz

code:

`def speak(text):
    speaker = gTTS(text = text, lang = "en")
    speaker.save("Voice.mp3")
    playsound.playsound("voice.mp3")`

![image](https://user-images.githubusercontent.com/58878720/105869642-70ef9680-601d-11eb-97de-b9e581e0a164.png)

**Comment by themdrizwanansari** [comment]
@fasterastv2, do you have read access to the location where Voice.mp3 is being saved?

**Comment by SG-SakethGupta** [comment]
@themdrizwanansari Just realise that the problem isnt with playing sound.

The actual problem is with the saving part. The line `speaker.save("Voice.mp3")` causes the problem

**Comment by venzen** [comment]
Here is a solution to play gTTS sound directly using `playsound` and `NamedTemporaryFile`:

```python
from gtts import gTTS
from tempfile import NamedTemporaryFile
from playsound import playsound

def speak(txt, lang='en'):
    gTTS(text=txt, lang=lang).write_to_fp(voice := NamedTemporaryFile())
    playsound(voice.name)
    voice.close()

txt = "Spank me, Daddy"
speak(txt)
```
Works in Python 3.9 and negates the need to save output to mp3.

**Comment by chajiuqqq** [comment]
@venzen ,thanks for your solution, only it works for me! 

```
from gtts import gTTS
from tempfile import NamedTemporaryFile
from playsound import playsound

def speak(txt, lang='en'):
    gTTS(text=txt, lang=lang).write_to_fp(voice := NamedTemporaryFile())
    playsound(voice.name)
    voice.close()

txt = "Spank me, Daddy"
speak(txt)
```

**Comment by GrimPixel** [comment]
> Here is a solution to play gTTS sound directly using `playsound` and `NamedTemporaryFile`:
> 
> ```python
> from gtts import gTTS
> from tempfile import NamedTemporaryFile
> from playsound import playsound
> 
> def speak(txt, lang='en'):
>     gTTS(text=txt, lang=lang).write_to_fp(voice := NamedTemporaryFile())
>     playsound(voice.name)
>     voice.close()
> 
> txt = "Spank me, Daddy"
> speak(txt)
> ```
> 
> Works in Python 3.9 and negates the need to save output to mp3.

Ohh, my hands hurt.

**Comment by EyesOfSiva** [comment]
tank you @venzen and @GrimPixel that worked like a charm

**Comment by Vuizur** [comment]
I tried pyglet, pygame and the tempfile + playsound solution, but they all failed due to various bugs (Windows 10 Python 3.10). Audio play doesn't seem to be a very stable area, seriously, or my PC is broken.
 
Then I tried this solution:
```python
from io import BytesIO
from pydub import AudioSegment
from pydub.playback import play

def say(text, lang = "en"):
    tts = gTTS(text=text, lang=lang)
    fp = BytesIO()
    tts.write_to_fp(fp)
    fp.seek(0)
    song = AudioSegment.from_file(fp, format="mp3")
    play(song)
```
which in theory should be efficient because it avoids writing the file on disk. Or so I thought, but it ended up being a few hundred ms slower than the naive solution which simply writes the file to disk. (I have an SSD though.) For a short text (5 words) the solution below is 200 - 300 ms faster.

```python
from playsound import playsound

def speak_gtts(words: str, lang: str="en"):
    filename = datetime.now().strftime("%Y-%m-%d_%H-%M-%S") + ".mp3"
    tts = gTTS(words, lang=lang)
    tts.save(filename)
    playsound(filename)
    os.remove(filename)
```

**Comment by NinjaSkibidigyat** [comment]
```from gtts import gTTS
from tempfile import NamedTemporaryFile
from playsound import playsound

def speak(txt, lang='en'):
    gTTS(text=txt, lang=lang).write_to_fp(voice := NamedTemporaryFile())
    playsound(voice.name)
    voice.close()

txt = "wow"
speak(txt)```

playsound is relying on another python subprocess. Please use `pip install pygobject` if you want playsound to run more efficiently.`
pip install pygobject
```Defaulting to user installation because normal site-packages is not writeable
Collecting pyobject
  Downloading pyobject-1.1.3.tar.gz (18 kB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [8 lines of output]
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/tmp/pip-install-1xdtuxk1/pyobject_dddd7e8ca7a54e6ca5ba8e5b72577132/setup.py", line 8, in <module>
          long_desc=open("README.rst").read()
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
        File "<frozen codecs>", line 322, in decode
      UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb8 in position 13: invalid start byte
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

**Comment by aadlinux358** [comment]
import io

from gtts import gTTS
from pydub import AudioSegment
from pydub.playback import play


hello = gTTS('hello')
fd = io.BytesIO()
hello.write_to_fp(fd)

fd.seek(0)

song = AudioSegment.from_mp3(fd)

play(song)

**Comment by BBC-Esq** [comment]
Loved the repository - thanks!  Here's my contribution using sounddevice and soundfile:

<details>
<summary>SAMPLE SCRIPT</summary>

```
# pip install gTTS sounddevice soundfile numpy

from gtts import gTTS
import sounddevice as sd
import soundfile as sf
import io
from gtts.tokenizer import pre_processors, tokenizer_cases, Tokenizer

def preprocess_text(text):
    text = pre_processors.abbreviations(text)
    text = pre_processors.end_of_line(text)
    text = pre_processors.tone_marks(text)
    return text

def tokenize_and_minimize(text):
    tokenizer = Tokenizer([
        tokenizer_cases.tone_marks,
        tokenizer_cases.period_comma,
        tokenizer_cases.colon,
        tokenizer_cases.other_punctuation
    ])
    
    tokens = tokenizer.run(text)
    
    minimized_tokens = []
    for token in tokens:
        if not token.strip():
            continue
        while len(token) > 100:
            split_index = token.rfind(' ', 0, 100)
            if split_index == -1:
                split_index = 100
            minimized_tokens.append(token[:split_index])
            token = token[split_index:].lstrip()
        minimized_tokens.append(token)
    
    return minimized_tokens

text = "Hello, how are you doing today? I'm doing well thank you for asking. Dr. Smith Jr. will see you now. This is a very-long-hyphenated-word. Let's add some tone marks: Wow! Amazing? Great!"

processed_text = preprocess_text(text)
print("Processed text:", processed_text)

tokens = tokenize_and_minimize(processed_text)
print("Tokens:", tokens)

fp = io.BytesIO()

# Process each token and concatenate the audio
for token in tokens:
    if token.strip():
        print(f"Processing token: '{token}'")
        # Default: English
        tts = gTTS(text=token, lang='en', slow=False, tld='com.ng')
        # For Spanish:
        # tts = gTTS(text=token, lang='es', slow=False)
        # For French:
        # tts = gTTS(text=token, lang='fr', slow=False)
        # For German:
        # tts = gTTS(text=token, lang='de', slow=False)
        # For Japanese:
        # tts = gTTS(text=token, lang='ja', slow=False)
        tts.write_to_fp(fp)

fp.seek(0)

# Read the audio data using soundfile
data, samplerate = sf.read(fp)

# Play the audio
sd.play(data, samplerate)
sd.wait()
```
</summary>

**Comment by adronstar** [comment]
```
import os

volume = 100
text = "No Need To Thank Me"

os.system(f'gtts-cli "{text}" | mpg123 - -g {volume}')
```
